File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userProfileUpdate/UserProfileUpdateHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userProfileUpdate

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userProfileUpdate.UserProfileUpdateApi
import features.userProfileUpdate.models.UserProfileModel

import features.userProfileUpdate.serializeUserProfileUpdateState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import features.userProfileUpdate.models.UpdateUserProfileRequest

class UserProfileUpdateHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/update_user_profile"

    override val eventChannel: String
        get() = "customer_sdk/events/update_user_profile"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return updateUserProfile(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun updateUserProfile(call: MethodCall): String? {
            val result = UserProfileUpdateApi().updateUserProfile(
                overrideHeaders = call.argument<Map<String, String>>("overrideHeaders") as Map<String, String>?,
                updateUserProfileRequest = Json.decodeFromString(call.argument<String>("updateUserProfileRequest") as String)
            ).toJsonResult<UserProfileModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserProfileUpdateApi().userProfileUpdateState.serializeUserProfileUpdateState().observe { result ->
            pushSerializedResultToEventSink<UserProfileUpdateApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/userProfileUpdate/UserProfileUpdateHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserProfileUpdateHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/update_user_profile"

    var eventChannel: String = "customer_sdk/events/update_user_profile"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserProfileUpdateHandler.updateUserProfile(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func updateUserProfile(call: FlutterMethodCall) async -> String? {


        
    let args = call.arguments
        do {

        var overrideHeaders: [String: String]?

        if let myArgs = args as? [String: Any] {
            overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
        }
        else {
            overrideHeaders = [:]
        }

        let result = try await UserProfileUpdateApi().updateUserProfile(overrideHeaders: overrideHeaders, methods: call.arguments as! [String : Any]).serializeUserProfileUpdateApiResult()

            switch result {
                case let result as SerializedResultSuccess:
                    return result.data
                case let result as SerializedResultUnauthorizedError:
                    return result.error
                case let result as SerializedResultUnknownError:
                    return result.error
                default:
                    return nil
            }

        }
        catch {
            return nil
        }


        


        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserProfileUpdateApi().userProfileUpdateState
            .serializeUserProfileUpdateState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
